Allow reusing specs built with compilers not in config

alalazo opened the pull request:
This PR allows to reuse specs that were built with compilers that **are not** in the current configuration. This means that specs from buildcaches don't need to have a matching compiler locally to be reused. If a node needs to be built, only available compilers will be considered as candidates.

Below most of the specs are reused and have `%gcc@=9.4.0` which is not in the configuration. The root node is built and has `%gcc@=11.1.0`:
![Screenshot from 2024-04-05 22-56-46](https://github.com/spack/spack/assets/4199709/85b39473-d958-4cd7-9cab-a939d8d80762)

alalazo committed:
Allow reusing non-local compilers

This commit allows reusing binaries with compilers that
don't match any in the local configuration.

alalazo committed:
Inject a Configuration object when retrieving compilers

This limits the implicit use of globals, and can
be leveraged to have simpler unit-tests.

alalazo committed:
Always monkeypatch cpu.host()

This is needed if we want to simulate x86_64
preferences on aarch64 host when running unit
tests

alalazo committed:
Add unit test for the compiler parser

haampie commented on lib/spack/spack/build_environment.py:
```diff
@@ -583,10 +584,13 @@ def set_package_py_globals(pkg, context: Context = Context.BUILD):
-    module.spack_cxx = os.path.join(link_dir, pkg.compiler.link_paths["cxx"])
-    module.spack_f77 = os.path.join(link_dir, pkg.compiler.link_paths["f77"])
-    module.spack_fc = os.path.join(link_dir, pkg.compiler.link_paths["fc"])
+    try:
```
This looks a bit too generic. Can't you catch a specific error? Maybe refactor to `compiler = get_compiler(pkg)` and then `if compiler: module.spack_cc = ...` etc?

alalazo replied:
Per DM, we need to check how to deal with `mpi` implementations, and similar specs that have a real run dependency on the compiler.

alalazo replied:
https://github.com/spack/spack/pull/43539/commits/4d117a574922879fc2ca834308c7e4ee18fca993 and https://github.com/spack/spack/pull/43539/commits/acca63bf2cd035cb9808a86128a8299fc419f01a

haampie commented on lib/spack/spack/test/concretize.py:
```diff
@@ -2360,6 +2362,26 @@ def test_select_lower_priority_package_from_repository_stack(
+    @pytest.mark.only_clingo("Old concretizer cannot reuse")
+    def test_reuse_specs_from_non_available_compilers(self, mutable_config, mutable_database):
+        """Tests that we can reuse specs with compilers that are not configured locally."""
+        # All the specs in the mutable DB have been compiled with %gcc@=10.2.1
```
maybe also assert  this? `mutable_database.query_local(...)`?

alalazo replied:
https://github.com/spack/spack/pull/43539/commits/d4bf4876aed5a801d9dc8b3ca3a74dacb7b9cfba

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2840,6 +2797,87 @@ def reject_requirement_constraint(
         return False
 
 
+class CompilerParser:
```
Parser? :thinking:

alalazo replied:
I'd leave this as is, at least here. There is also a `RequirementParser` - open to change both names in a following Pr.

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2840,6 +2797,87 @@ def reject_requirement_constraint(
+        result = sorted(self.compilers, key=lambda x: (x.spec.name, x.spec.version), reverse=True)
+        # Then stable sort to prefer available compilers and account for preferences
+        ppk = spack.package_prefs.PackagePrefs("all", "compiler", all=False)
+        result = sorted(result, key=lambda x: (not x.available, ppk(x.spec)))
```
avoid a copy `result.sort(...)`

alalazo replied:
https://github.com/spack/spack/pull/43539/commits/780bfccc0ddc2e14971c5ae4f3ddff8b7538ecf9

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2840,6 +2797,87 @@ def reject_requirement_constraint(
+
+            version = s.compiler.versions.concrete
+
+            if not version or any(item.spec.satisfies(s.compiler) for item in self.compilers):
```
Isn't s.compiler the same object `id` almost always? Given that a compiler is almost always constant this sounds inefficient.

Alternatively the linear scan could be avoided maybe with `s.compiler in set_of_seen_compiler_specs`?

Edit: not the same `id`, but equal. I would use a set of CompilerSpec objects to avoid the linear scan. `satisfies` subset semantics makes not sense for concrete specs anyways.

haampie replied:
Also `not version` should in principle be implied by `s.concrete`

alalazo replied:
> satisfies subset semantics makes not sense for concrete specs anyways.

Is this based on a misunderstanding:
```python
for s in traverse.traverse_nodes(input_specs):
    # we don't need to validate compilers for already-built specs
    if s.concrete or not s.compiler:
        continue
```
? If `s` is concrete we won't reach that code, and move on to the next iteration. Note that this is run for input specs, not on reusable concrete specs.

alalazo replied:
I'm also shortening the check in the other method.

alalazo committed:
Use a more specific exception when a compiler is not available

alalazo committed:
Assert that mock specs are compiled with gcc@10.2.1

alalazo committed:
Avoid a copy on return value

alalazo committed:
Shorten the checks for reused specs

alalazo committed:
Fix mpi packages to account for possibly missing compilers

This would happen if the mpi was installed from
a buildcache, and the compiler is not available.

alalazo commented:
@spackbot run pipeline
